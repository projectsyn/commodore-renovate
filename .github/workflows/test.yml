name: Test
on:
  pull_request:
    branches:
      - master

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Lint
        run: |
          yarn lint
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '18'
          cache: yarn
      - name: Installing dependencies
        run: yarn install --frozen-lockfile
      - name: Setup Python
        uses: actions/setup-python@v4
        with:
          python-version: '3.9'
      - name: Cache pip
        uses: actions/cache@v3
        with:
          # This path is specific to Ubuntu
          path: ~/.cache/pip
          # Look to see if there is a cache hit for the corresponding requirements file
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
      - name: Install dependencies
        run: pip install -r requirements.txt
      - name: Test
        run: |
          yarn test

  docker-build:
    runs-on: ubuntu-latest
    env:
      IMAGE: ${{ github.repository }}
    steps:
      - uses: actions/checkout@v4
      - name: Get changed files
        id: changed-files
        uses: tj-actions/changed-files@v39
      - name: Build
        if: contains(steps.changed-files.outputs.modified_files, 'Dockerfile')
        uses: docker/build-push-action@v4
        with:
          push: false
          tags: 'ghcr.io/${{ env.IMAGE }}:test'
