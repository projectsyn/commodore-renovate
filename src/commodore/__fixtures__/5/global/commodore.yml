classes:
  - global.params
  - global.${facts:distribution}
  - ${cluster:tenant}.${cluster:name}
