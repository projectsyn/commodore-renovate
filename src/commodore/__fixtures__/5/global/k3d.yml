parameters:
  components:
    argocd:
      url: https://github.com/projectsyn/component-argocd-k3d.git
