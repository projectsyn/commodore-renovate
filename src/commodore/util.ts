import clone from 'just-clone';
import yaml from 'js-yaml';
import Git from 'simple-git';

import { parse } from 'path';
import { mkdir, readFile, writeFile } from 'fs/promises';
import { URL } from 'url';

import { logger } from 'renovate/dist/logger';
import { getGlobalConfig } from 'renovate/dist/config/global';
import { simpleGitConfig } from 'renovate/dist/util/git/config';

import type { Facts, RepoConfig } from './types';

export function cacheDir(): string {
  let cacheDir: string | undefined = getGlobalConfig().cacheDir;
  if (cacheDir === undefined) {
    cacheDir = '/tmp/renovate';
  }
  return cacheDir;
}

/* Prune object, removing any null or undefined keys */
export function pruneObject(obj: Facts) {
  return Object.entries(obj)
    .filter(([_, v]) => v !== undefined && v !== null)
    .reduce((main, [key, value]) => ({ ...main, [key]: value }), {});
}

/* Render object as YAML and write to file */
export async function writeYamlFile(
  filePath: string,
  obj: object
): Promise<void> {
  const objYaml: string = yaml.dump(obj);
  return writeFile(filePath, objYaml);
}

/* Extract distribution/cloud/region from filename */
export function parseFileName(
  fileName: string,
  distributionRegex: RegExp,
  cloudRegionRegex: RegExp,
  ignoreValues: Array<string>
): Facts {
  const distA = fileName.match(distributionRegex);
  const cloudA = fileName.match(cloudRegionRegex);

  const dist = distA ? distA.groups : null;
  const cloud = cloudA ? cloudA.groups : null;

  let facts: Facts = {
    distribution: null,
    cloud: null,
    region: null,
  };
  if (dist !== undefined && dist !== null) {
    facts.distribution = dist.distribution;
    facts.cloud = dist.cloud ? dist.cloud : null;
  }
  if (cloud !== undefined && cloud !== null) {
    facts.cloud = cloud.cloud;
    facts.region = cloud.region ? cloud.region : null;
  }

  if (
    facts.distribution !== null &&
    ignoreValues.includes(facts.distribution)
  ) {
    facts.distribution = null;
  }
  if (facts.cloud !== null && ignoreValues.includes(facts.cloud)) {
    facts.cloud = null;
  }
  if (facts.region !== null && ignoreValues.includes(facts.region)) {
    facts.region = null;
  }

  return facts;
}

function isObject(item: any) {
  return item && typeof item === 'object' && !Array.isArray(item);
}

// Adapted from https://stackoverflow.com/a/37164538
export function mergeConfig(base: any, config: any): any {
  let output = clone(base);
  if (isObject(base) && isObject(config)) {
    Object.keys(config).forEach((key) => {
      if (isObject(config[key])) {
        if (!(key in base)) Object.assign(output, { [key]: config[key] });
        else output[key] = mergeConfig(base[key], config[key]);
      } else if (Array.isArray(config[key])) {
        if (!(key in base)) {
          Object.assign(output, { [key]: Array.from(config[key]) });
        } else if (Array.isArray(output[key])) {
          output[key].push(...config[key]);
        } else {
        }
      } else {
        Object.assign(output, { [key]: config[key] });
      }
    });
  }
  return output;
}

export var globalRepos: Map<string, RepoConfig> = new Map();

export function globalRepoDir(globalRepoURL: string): string {
  let repoUrl = new URL(globalRepoURL);
  let repoPathInfo = parse(repoUrl.pathname);
  let repoPath = `${repoUrl.host}${repoPathInfo.dir}/${repoPathInfo.name}`;
  return cacheDir() + `/global-repos/${repoPath}`;
}

export async function cloneGlobalRepo(config: any): Promise<RepoConfig> {
  if (globalRepos.has(config.globalRepoURL)) {
    return globalRepos.get(config.globalRepoURL) as RepoConfig;
  }

  logger.info(
    {
      globalRepoURL: config.globalRepoURL,
      tenantId: config.tenantId,
    },
    'Global repo for this tenant is not initialized yet, cloning it'
  );
  const dir: string = globalRepoDir(config.globalRepoURL);
  await mkdir(dir, { recursive: true });
  // TODO(sg): support Git-https?
  const git: any = Git(dir, simpleGitConfig());
  await git.clone(config.globalRepoURL, '.');
  const globalExtraConfig = await parseGlobalRepoConfig(dir);
  const rc: RepoConfig = {
    dir: dir,
    extraConfig: globalExtraConfig,
  } as RepoConfig;
  globalRepos.set(config.globalRepoURL, rc);
  return rc;
}

export async function parseGlobalRepoConfig(globalDir: string): Promise<any> {
  const globalDirConfigFile: string = globalDir + '/renovate.json';
  let globalDirConfig: any = {};
  try {
    const globalConfigStr: string = (
      await readFile(globalDirConfigFile)
    ).toString();
    globalDirConfig = JSON.parse(globalConfigStr) as any;
  } catch (e: any) {
    logger.info(
      {
        globalDir: globalDir,
        exception: e,
      },
      'Exception while reading global repo config'
    );
  }
  if (
    globalDirConfig.commodore != null &&
    globalDirConfig.commodore.extraConfig != null
  ) {
    try {
      const globalExtraConfigStr = (
        await readFile(globalDir + '/' + globalDirConfig.commodore.extraConfig)
      ).toString();
      return JSON.parse(globalExtraConfigStr) as any;
    } catch (e: any) {
      logger.info(
        {
          globalDir: globalDir,
          extraConfig: globalDirConfig.commodore.extraConfig,
          exception: e,
        },
        'Exception while reading global repo extraConfig'
      );
    }
  }
  return {};
}
