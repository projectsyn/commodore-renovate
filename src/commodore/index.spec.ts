import { beforeAll, afterAll, describe, expect, it } from '@jest/globals';
import { copyFile, mkdir, readdir } from 'fs/promises';
import { rmSync } from 'fs';

import Git from 'simple-git';

import { GlobalConfig } from 'renovate/dist/config/global';

import { getFixturePath, getLoggerErrors, loadFixture } from '../test/util';
import { defaultConfig, extractPackageFile } from './index';

import nock from 'nock';

import { clearProblems } from 'renovate/dist/logger';

const params1 = loadFixture('1/params.yml');
const kube2 = loadFixture('2/kubernetes.yml');
const invalid3 = loadFixture('3/params.yml');
const pin4 = loadFixture('4/pins.yml');
const tenant1 = loadFixture('5/tenant/c-foo.yml');
const tenant2 = loadFixture('5/tenant/c-foo-2.yml');
const tenant3 = loadFixture('5/tenant/c-foo-3.yml');
const tenant4 = loadFixture('5/tenant/c-foo-4.yml');
const tenantTest = loadFixture('5/tenant/test.yml');

function setGlobalConfig(fixtureId: string, isTenantFixture: boolean): void {
  GlobalConfig.get().localDir =
    getFixturePath(fixtureId) + (isTenantFixture ? '/tenant' : '');
  GlobalConfig.get().cacheDir = '/tmp/renovate';
}

async function setupGlobalRepo(
  fixtureId: string,
  targetId: string
): Promise<string> {
  const repoDir = `/tmp/renovate/${targetId}/commodore-defaults`;
  await mkdir(repoDir, { recursive: true });
  const git = Git({ baseDir: repoDir });
  await git.init();
  await git.addConfig('user.name', 'Commodore-Renovate Unit Tests');
  await git.addConfig('user.email', 'ahoy@syn.tools');
  const sourceDir = getFixturePath(fixtureId);
  const repoContents = await readdir(sourceDir);
  repoContents.forEach((file) => {
    copyFile(sourceDir + '/' + file, repoDir + '/' + file);
    git.add(file);
  });
  await git.commit('initial commit');
  return repoDir;
}

function setupNock(
  clusterId: string,
  statusCode: number,
  reason: string,
  dynamicFacts: boolean
): nock.Scope {
  let reply200: any = {
    id: clusterId,
    tenant: 't-bar',
    displayName: 'Foo cluster',
    facts: {
      cloud: 'none',
      distribution: 'k3d',
      'lieutenant-instance': 'example-com',
    },
    gitRepo: {
      deployKey: 'ssh-rsa AAAAB...',
      hostKeys: 'git.example.com ssh-rsa AAAAB...',
      type: 'auto',
      url: 'ssh://git@git.example.com/cluster-catalogs/c-cluster-id-1234.git',
    },
  };
  if (dynamicFacts) {
    // Change facts when testing response without dynamic facts to avoid
    // reusing cached response, since we don't consider dynamic facts when
    // caching commodore results at the moment.
    reply200.facts.cloud = 'local';
    reply200.dynamicFacts = {
      fact: '1',
      fact2: '2',
    };
  }
  const scope = nock('https://lieutenant.example.com', {
    reqheaders: {
      authorization: 'Bearer mock-api-token',
    },
  }).get(`/clusters/${clusterId}`);
  if (statusCode == 200) {
    return scope.reply(statusCode, reply200);
  }
  return scope.reply(statusCode, { reason: reason });
}

beforeAll(() => {
  return mkdir('/tmp/renovate', { recursive: true });
});

beforeEach(() => {
  // clear logger
  clearProblems();
});

afterAll(() => {
  // Remove global repo clone created by the test
  rmSync('/tmp/renovate/global-repos/', { recursive: true });
});

describe('src/commodore/index', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      setGlobalConfig('1', false);
      return expect(
        extractPackageFile('nothing here', 'no.yml', defaultConfig)
      ).resolves.toBeNull();
    });
    it('extracts component and package versions', async () => {
      setGlobalConfig('2', false);
      const res = await extractPackageFile(
        params1,
        '1/params.yml',
        defaultConfig
      );
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(8);
      }
    });
    it('returns no component or package version for files without components or packages', () => {
      return expect(
        extractPackageFile(kube2, '2/kubernetes.yml', defaultConfig)
      ).resolves.toBeNull();
    });
    it('returns null for invalid yaml', () => {
      setGlobalConfig('3', false);
      return expect(
        extractPackageFile(invalid3, '3/params.yml', defaultConfig)
      ).resolves.toBeNull();
    });
    it('extracts component and package urls for version pins', async () => {
      setGlobalConfig('4', false);
      const res = await extractPackageFile(pin4, '4/pins.yml', defaultConfig);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }
    });
    it('extracts component and package urls for version pins in tenant repo', async () => {
      setGlobalConfig('5', true);
      const globalRepoDir: string = await setupGlobalRepo('5/global', '5');
      const config: any = { ...defaultConfig };
      config.tenantId = 't-bar';
      config.globalRepoURL = `file://${globalRepoDir}`;
      const res = await extractPackageFile(
        tenant1,
        '5/tenant/c-foo.yml',
        config
      );
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }

      // Clean up global repo for this test case
      rmSync('/tmp/renovate/5', { recursive: true });
    });
    it('uses cluster info returned by Lieutenant', async () => {
      setGlobalConfig('5', true);
      const globalRepoDir: string = await setupGlobalRepo('5/global', '6');
      const scope = setupNock('c-foo-2', 200, '', true);
      const config: any = { ...defaultConfig };
      config.lieutenantURL = 'https://lieutenant.example.com';
      config.lieutenantTokenEnvVar = 'LIEUTENANT_API_TOKEN';
      process.env.LIEUTENANT_API_TOKEN = 'mock-api-token';
      config.tenantId = 't-bar';
      config.globalRepoURL = `file://${globalRepoDir}`;
      config.dynamic_facts;
      const res = await extractPackageFile(
        tenant2,
        '5/tenant/c-foo-2.yml',
        config
      );
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }
      expect(scope.isDone()).toBe(true);

      // Clean up global repo for this test case
      rmSync('/tmp/renovate/6', { recursive: true });
    });
    it('proceeds without cluster info on 404', async () => {
      setGlobalConfig('5', true);
      const globalRepoDir: string = await setupGlobalRepo('5/global', '7');
      const scope = setupNock('c-foo-3', 404, 'cluster not found', false);
      const config: any = { ...defaultConfig };
      config.lieutenantURL = 'https://lieutenant.example.com';
      config.lieutenantTokenEnvVar = 'LIEUTENANT_API_TOKEN';
      process.env.LIEUTENANT_API_TOKEN = 'mock-api-token';
      config.tenantId = 't-bar';
      config.globalRepoURL = `file://${globalRepoDir}`;
      const res = await extractPackageFile(
        tenant3,
        '5/tenant/c-foo-3.yml',
        config
      );
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }
      expect(scope.isDone()).toBe(true);

      // Clean up global repo for this test case
      rmSync('/tmp/renovate/7', { recursive: true });
    });
    it('uses facts from factsMap', async () => {
      const globalRepoDir: string = await setupGlobalRepo('5/global', '8');
      const config: any = { ...defaultConfig };
      config.tenantId = 't-bar';
      config.globalRepoURL = `file://${globalRepoDir}`;
      config.extraConfig = 'renovate.commodore.json';
      const res = await extractPackageFile(
        tenantTest,
        '5/tenant/test.yml',
        config
      );
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }

      // Clean up global repo for this test case
      rmSync('/tmp/renovate/8', { recursive: true });
    });
    it('generates valid Commodore parameters for clusters with no dynamic facts', async () => {
      setGlobalConfig('5', true);
      const globalRepoDir: string = await setupGlobalRepo('5/global', '9');
      const scope = setupNock('c-foo-4', 200, '', false);
      const config: any = { ...defaultConfig };
      config.lieutenantURL = 'https://lieutenant.example.com';
      config.lieutenantTokenEnvVar = 'LIEUTENANT_API_TOKEN';
      process.env.LIEUTENANT_API_TOKEN = 'mock-api-token';
      config.tenantId = 't-bar';
      config.globalRepoURL = `file://${globalRepoDir}`;
      const res = await extractPackageFile(
        tenant4,
        '5/tenant/c-foo-4.yml',
        config
      );
      const errors = getLoggerErrors();
      if (errors.length > 0) {
        console.log(errors);
      }
      expect(errors.length).toBe(0);
      expect(res).not.toBeNull();
      if (res) {
        const deps = res.deps;
        expect(deps).toMatchSnapshot();
        expect(deps).toHaveLength(3);
      }
      expect(scope.isDone()).toBe(true);

      // Clean up global repo for this test case
      rmSync('/tmp/renovate/9', { recursive: true });
    });
  });
});
