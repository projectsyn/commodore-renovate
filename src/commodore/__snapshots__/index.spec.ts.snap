// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/commodore/index extractPackageFile() extracts component urls for version pins 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 4/pins.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 4/pins.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() extracts component urls for version pins in tenant repo 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 5/tenant/c-foo.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 5/tenant/c-foo.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() extracts component versions 1`] = `
Array [
  Object {
    "currentValue": "v1.1.0",
    "depName": "appuio-cloud in 1/params.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.0.0",
    "depName": "appuio-cloud2 in 1/params.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.0.0",
    "depName": "argocd in 1/params.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd.git",
  },
  Object {
    "currentValue": "v2.1.0",
    "depName": "backup-k8up in 1/params.yml",
    "lookupName": "https://github.com/projectsyn/component-backup-k8up.git",
  },
  Object {
    "currentValue": "v1.0.0",
    "depName": "kyverno in 1/params.yml",
    "lookupName": "https://github.com/projectsyn/component-kyverno.git",
  },
  Object {
    "currentValue": "v2.2.0",
    "depName": "lieutenant in 1/params.yml",
    "lookupName": "https://github.com/projectsyn/component-lieutenant.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() generates valid Commodore parameters for clusters with no dynamic facts 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 5/tenant/c-foo-4.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 5/tenant/c-foo-4.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd-k3d.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() proceeds without cluster info on 404 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 5/tenant/c-foo-3.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 5/tenant/c-foo-3.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() uses cluster info returned by Lieutenant 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 5/tenant/c-foo-2.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 5/tenant/c-foo-2.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd-k3d.git",
  },
]
`;

exports[`src/commodore/index extractPackageFile() uses facts from factsMap 1`] = `
Array [
  Object {
    "currentValue": "v3.0.0",
    "depName": "appuio-cloud in 5/tenant/test.yml",
    "lookupName": "https://github.com/appuio/component-appuio-cloud.git",
  },
  Object {
    "currentValue": "v1.2.3",
    "depName": "argocd in 5/tenant/test.yml",
    "lookupName": "https://github.com/projectsyn/component-argocd-k3d.git",
  },
]
`;
